### PR TITLE
Annotations, Actions, and more

### PR DESCRIPTION
**Pull request by reknih**
_it's spec time, baby_

This PR includes annotations (insert links and tell your co-authors why exactly their work is shite), actions (make your PDF do things that it clearly should not be able to do), more stuff on the document catalogue (setting RTL as the document direction and more), transitions (who needs PowerPoint anyways) and an example to boost.

It introduces two new data types: `Date`s and `TextStr` (UTF16 strings).

### CODE REVIEW

File: src/structure.rs [hunk 1]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+        const LOCKED = 1 << 7;
+        /// Invert the interpretation of the `no_view` flag for certain events.
+        /// (1.5+)
+        const TOGGLE_NO_VIEW = 1 << 8;
+        /// Do not allow content changes. (1.7+)
+        const LOCKED_CONTENTS = 1 << 9;
+    }
+}
+
+/// Highlighting effect.
```
Comment (laurmaedje) [inline src/structure.rs:681]: this could be more descriptive

File: src/structure.rs [hunk 2]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+impl<'a> FileSpec<'a> {
+    pub(crate) fn new(obj: Obj<'a>) -> Self {
+        let mut dict = obj.dict();
+        dict.pair(Name(b"Type"), Name(b"Filespec"));
+        Self { dict }
+    }
+
+    /// Write the `/FS` attribute to set the file system this entry relates to.
+    /// If you set the `system` argument to `Name(b"URL")`, this becomes a URL
+    /// Specification.
```
Comment (laurmaedje) [inline src/structure.rs:721]:
```suggestion
    /// If you set the `system` argument to `Name(b"URL")`, this becomes an URL
    /// specification.
```

File: src/structure.rs [hunk 3]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    pub fn file_system(&mut self, system: Name) -> &mut Self {
+        self.pair(Name(b"FS"), system);
+        self
+    }
+
+    /// Write the `/F` attribute to set the file path. Directories are indicated
+    /// by `/`, independant of the platform.
```
Comment (laurmaedje) [inline src/structure.rs:728]:
```suggestion
    /// by `/`, independent of the platform.
```

File: src/structure.rs [hunk 4]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    pub fn file(&mut self, path: Str) -> &mut Self {
+        self.pair(Name(b"F"), path);
+        self
+    }
+
+    /// Write the `/UF` attribute to set a Unicode-compatible path. Directories
+    /// are indicated by `/`, independant of the platform. (1.7+)
```
Comment (laurmaedje) [inline src/structure.rs:735]:
```suggestion
    /// are indicated by `/`, independent of the platform. (1.7+)
```

File: src/structure.rs [hunk 5]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    /// Write the `/Subj` attribute. This is the subject of the annotation.
+    /// (1.5+)
+    pub fn subject(&mut self, text: TextStr) -> &mut Self {
+        self.pair(Name(b"Subj"), text);
+        self
+    }
+
+    /// Start writing the `/BS` attribute. These are some more elaborate border
+    /// settings taking precedence over `/B` for some annotation types. (1.6+)
+    pub fn border_style(&mut self) -> BorderStyle<'_> {
```
Comment (laurmaedje) [inline src/structure.rs:508]: this could live next to `border`. also, I think it's 1.2+ and not 1.6+.

File: src/structure.rs [hunk 6]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+impl<'a> Destination<'a> {
+    pub(crate) fn start(obj: Obj<'a>, page: Ref) -> Self {
+        let mut array = obj.array();
+        array.item(page);
+        Self { array }
+    }
+
+    /// Write the `/XYZ` command which skips to the specified coordinated. Leave
+    /// `zoom` on `0` if you do not want to change the zoom level.
+    pub fn xyz(mut self, left: f32, top: f32, zoom: f32) {
```
Comment (laurmaedje) [inline src/structure.rs:833]: could `zoom` just be an `Option<f32>`?

File: src/structure.rs [hunk 7]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    dict: Dict<'a, IndirectGuard>,
+}
+
+impl<'a> Destinations<'a> {
+    pub(crate) fn start(obj: Obj<'a, IndirectGuard>) -> Self {
+        Self { dict: obj.dict() }
+    }
+
+    /// Start adding another named destination.
+    pub fn add(&mut self, name: Name, page: Ref) -> Destination<'_> {
```
Comment (laurmaedje) [inline src/structure.rs:906]: also `push`?

File: src/structure.rs [hunk 8]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+        }
+
+        self.pair(Name(b"Count"), items);
+        self
+    }
+}
+
+deref!('a, Outline<'a> => Dict<'a, IndirectGuard>, dict);
+
+/// Writer for an _outline dictionary_.
```
Comment (laurmaedje) [inline src/structure.rs:956]:
```suggestion
/// Writer for an _outline item dictionary_.
```

File: src/structure.rs [hunk 9]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    }
+
+    /// Write the `/Dest` attribute to set the destination of this
+    /// outline item to a named destination.
+    pub fn dest_name(&mut self, name: Name) -> &mut Self {
+        self.pair(Name(b"Dest"), name);
+        self
+    }
+
+    /// Write the `/C` attribute using a RGB color. This sets the color in which
```
Comment (laurmaedje) [inline src/structure.rs:1033]:
```suggestion
    /// Write the `/C` attribute using an RGB color. This sets the color in which
```

File: src/structure.rs [hunk 10]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    }
+
+    /// Write the `/C` attribute using a grayscale color. This sets the
+    /// annotations background color and its popup title bar color. (1.1+)
+    pub fn color_gray(&mut self, gray: f32) -> &mut Self {
+        self.key(Name(b"C")).array().typed().item(gray);
+        self
+    }
+
+    /// Write the `/C` attribute using a RGB color. This sets the annotations
```
Comment (laurmaedje) [inline src/structure.rs:464]:
```suggestion
    /// Write the `/C` attribute using an RGB color. This sets the annotations
```

File: src/structure.rs [hunk 11]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+            Self::Squiggly => Name(b"Squiggly"),
+            Self::StrikeOut => Name(b"StrikeOut"),
+            Self::FileAttachment => Name(b"FileAttachment"),
+        }
+    }
+}
+
+/// A name that sets the icon for the annotation.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum AnnotationName {
```
Comment (laurmaedje) [inline src/structure.rs:604]: maybe?
```suggestion
/// Possible icons for an annotation.
#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
pub enum AnnotationIcon {
```

File: src/lib.rs [hunk 12]
```diff
@@ -99,13 +100,22 @@ pub mod writers {
     use super::*;
     pub use content::{ImageStream, Path, Text};
     pub use font::{CidFont, CmapStream, FontDescriptor, Type0Font, Type1Font, Widths};
-    pub use structure::{Catalog, Page, Pages, Resources};
+    pub use interactive::{Action, Transition};
+    pub use structure::{
+        Annotation, Annotations, Catalog, Destination, Destinations, FileSpec, Outline,
+        OutlineItem, Page, Pages, Resources, ViewerPreferences,
```
Comment (laurmaedje) [inline src/lib.rs:106]: `BorderStyle` is missing

File: src/structure.rs [hunk 13]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+            Self::UseNone => Name(b"UseNone"),
+            Self::UseOutlines => Name(b"UseOutlines"),
+            Self::UseThumbs => Name(b"UseThumbs"),
+            Self::FullScreen => Name(b"FullScreen"),
+        }
+    }
+}
+
+/// Predominant reading order of text. Used to aid the viewer with the spacial
+/// ordering in which to display pages.
```
Comment (laurmaedje) [inline src/structure.rs:333]:
```suggestion
/// Predominant reading order of text.
///
/// Used to aid the viewer with the spacial ordering in which to display pages.
```
looks better in docs

File: src/structure.rs [hunk 14]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+            Self::Solid => Name(b"S"),
+            Self::Dashed => Name(b"D"),
+            Self::Beveled => Name(b"B"),
+            Self::Inset => Name(b"I"),
+            Self::Underline => Name(b"U"),
+        }
+    }
+}
+
+/// Writer for the _destination array_.
```
Comment (laurmaedje) [inline src/structure.rs:819]:
```suggestion
/// Writer for a _destination array_.
///
/// This struct is created by [`Destinations::push`].
```

File: src/structure.rs [hunk 15]
```diff
@@ -128,3 +268,790 @@ impl<'a> Resources<'a> {
...
+    /// the screen height and skips to the specified offset. (1.1+)
+    pub fn fit_bounding_box_vertical(mut self, left: f32) {
+        self.item(Name(b"FitBV"));
+        self.item(left);
+    }
+}
+
+deref!('a, Destination<'a> => Array<'a>, array);
+
+/// Writer for a _named destinations dictionary_.
```
Comment (laurmaedje) [inline src/structure.rs:893]: This whole struct + impl should be above `Destination`

**Review by laurmaedje — CHANGES_REQUESTED** [review]

File: examples/hello.rs [hunk 16]
```diff
@@ -24,20 +27,45 @@ fn main() -> std::io::Result<()> {
...
     // is only a font that we name "F1" (the specific name doesn't matter).
-    writer
-        .page(page_id)
-        .parent(page_tree_id)
-        .media_box(Rect::new(0.0, 0.0, 595.0, 842.0))
-        .contents(text_id)
-        .resources()
-        .fonts()
-        .pair(font_name, font_id);
+    page.resources().fonts().pair(font_name, font_id);
```
[the remark this answers is not in the document]
Comment (reknih) [reply examples/hello.rs:66]: This is actually the settings of resources / fonts, so it has to be a chain

[the remark this answers is not in the document]
Comment (laurmaedje) [reply examples/hello.rs:67]: oh, true

File: examples/hello.rs [hunk 17]
```diff
@@ -37,29 +37,30 @@ fn main() -> std::io::Result<()> {
...
     // We also create the annotations list here that allows us to have things
     // like links or comments on the page.
-    let mut annots = page.annotations();
+    let mut annotations = page.annotations();
 
     // Write a new annotation.
-    let mut annot = annots.add();
+    let mut annotation = annotations.push();
 
     // Write the type, area, and action for this annotation.
```
Comment (laurmaedje) [inline examples/hello.rs:45]: I feel like this comments lists a random portion of the things that are actually configured.

**Review by laurmaedje — APPROVED** [review]

**Comment by laurmaedje** [comment]
Thank you!

**Pull request merged**